### PR TITLE
Add query running functions to the query service client

### PR DESCRIPTION
**Pull request by samredai**
### Summary

This adds functions to the query service client for submitting queries and retrieving them (when running async).

### Test Plan

N/A

- [x] PR has an associated issue: #342 
- [x] `make check` passes
- [x] `make test` shows 100% unit test coverage

### Deployment Plan

N/A

### CODE REVIEW

File: dj/service_clients.py [hunk 1]
```diff
@@ -82,3 +84,46 @@ def get_columns_for_table(
...
+                "catalog_name": catalog,
+                "async_": async_,
+            },
+        )
+        if response.status_code >= 400:
+            raise DJException(
+                message=f"Error response from query service: {response.text}",
+            )
+        query_info = response.json()
+        return QueryWithResults(**query_info)
```
Comment (agorajek) [inline dj/service_clients.py:114]: Let's make it `QueryResults` please? :) 

`QueryWithResults` feels a bit weird. It makes me want to ask for `QueryWithoutResults` as well.

**Review by agorajek — APPROVED** [review]
Looks good. Just one silly request in-line.

File: dj/service_clients.py (hunk 1 above)
Comment (samredai) [reply dj/service_clients.py:114]: Haha, every-time I see that I think of how much you're not a fan of that class name. 😂 I have a big refactor coming up to fully complete the move to the API and wiring up DJQS. Once that's in I'll remember to come back and update the class name here and in DJQS. 💯

Comment (betodealmeida) [reply dj/service_clients.py:114]: @agorajek but this is a query object with results! 🙂 

It has the query information (status, progress, submitted/executed query) and the results. It's almost identical to the `Query` payload, with the addition of the result, hence `QueryWithResults`.

https://github.com/DataJunction/dj/blob/110168faedd7d052f86107ff71c0c60dc648d93c/dj/models/query.py#L114

File: dj/service_clients.py [hunk 2]
```diff
@@ -82,3 +84,46 @@ def get_columns_for_table(
...
+
+    def get_query(
+        self,
+        query_id: str,
+    ) -> QueryWithResults:
+        """
+        Get a previously submitted query
+        """
+        response = self.requests_session.get(f"/queries/{query_id}/")
+        if response.status_code >= 400:
```
Comment (betodealmeida) [inline dj/service_clients.py:124]: Nit:

```suggestion
        if not response.ok:
```

**Review by betodealmeida — CHANGES_REQUESTED** [review]
@samredai I left a few comments, especially about the naked `DJException`. I'm going through the code base converting all of them to more specific exceptions, it would be nice if we could do the same here.

File: dj/service_clients.py (hunk 1 above)
Comment (betodealmeida) [reply dj/service_clients.py:114]: And for context, the idea with `QueryWithResults` is that it simplifies the query endpoint. You can a single endpoint where you can create queries (doing a `POST`), check for their status (`GET`) and fetch results (also `GET`).

File: dj/service_clients.py [hunk 3]
```diff
@@ -82,3 +84,46 @@ def get_columns_for_table(
...
+            json={
+                "engine": engine,
+                "engine_version": engine_version,
+                "submitted_query": submitted_query,
+                "catalog_name": catalog,
+                "async_": async_,
+            },
+        )
+        if response.status_code >= 400:
+            raise DJException(
```
[the remark this answers is not in the document]
Comment (samredai) [reply dj/service_clients.py]: Great point! We should definitely make sure errors are as clear as possible 💯 . Updated the PR to use a `DJQueryServiceClientException`

File: dj/service_clients.py [hunk 4]
```diff
@@ -82,3 +84,46 @@ def get_columns_for_table(
...
+        query_id: str,
+    ) -> QueryWithResults:
+        """
+        Get a previously submitted query
+        """
+        response = self.requests_session.get(f"/queries/{query_id}/")
+        if response.status_code >= 400:
+            raise DJException(
+                message=f"Error response from query service: {response.text}",
+            )
```
[the remark this answers is not in the document]
Comment (samredai) [reply dj/service_clients.py]: Added a dbapi_exception to the `DJQueryServiceClientException`. Thanks @betodealmeida!
```py
class DJQueryServiceClientException(DJException):
    """
    Exception raised when the query service returns an error
    """

    dbapi_exception: DBAPIExceptions = "InterfaceError"
    http_status_code: int = 500
```

**Review by shangyian — APPROVED** [review]

File: dj/errors.py [hunk 5]
```diff
@@ -217,3 +217,12 @@ class DJInternalErrorException(DJException):
 
     dbapi_exception: DBAPIExceptions = "InternalError"
     http_status_code: int = 500
+
+
+class DJQueryServiceClientException(DJException):
```
Comment (shangyian) [inline dj/errors.py:222]: Nice, glad that we're going for more specific exceptions going forward.

File: dj/service_clients.py (hunk 1 above)
Comment (agorajek) [reply dj/service_clients.py:114]: hold-on-hmm

OK, I will not bring this subject again. If I do please add it to my performance review.

**Review by betodealmeida — APPROVED** [review]
Thanks for addressing the comments!

**Pull request merged**